packages/material-ui/src/ButtonBase/ButtonBase.js

```diff
@@ -89,16 +77,15 @@ class ButtonBase extends React.Component {
 
   handleContextMenu = createRippleHandler(this, 'ContextMenu', 'stop');
 
-  handleBlur = createRippleHandler(this, 'Blur', 'stop', () => {
-    clearTimeout(this.focusVisibleTimeout);
+  handleBlur = createRippleHandler(this, 'Blur', 'stop', event => {
     if (this.state.focusVisible) {
+      handleBlurVisible(event);
       this.setState({ focusVisible: false });
     }
   });
 
   componentDidMount() {
-    listenForFocusKeys(ownerWindow(this.getButtonNode()));
-
+    prepareFocusVisible(this.getButtonNode().ownerDocument);
```

Comment:
@eps1lon It does look like this is going to add a set of document listeners for each `ButtonBase` mount. Should be able to short-circuit within `focusVisible.js` `prepare` function by tracking in a variable whether or not the listeners have already been added to the document.

Reply:
Should we handle it?

Reply:
>  It does look like this is going to add a set of document listeners for each ButtonBase mount.

[addEventListener does not add duplicate listeners](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#Multiple_identical_event_listeners). Unless the focusVisible.js module is recreated for every ButtonBase instance they shouldn't be duplicated.

Reply:
Thanks for the education.